cfgmgr/coppmgr.cpp

```diff
@@ -250,7 +253,65 @@ void CoppMgr::mergeConfig(CoppCfg &init_cfg, CoppCfg &m_cfg, std::vector<std::st
         {
             std::vector<FieldValueTuple> cfg_fvs;
             cfgTable.get(i, cfg_fvs);
-            m_cfg[i] = cfg_fvs;
+            helper[i] = cfg_fvs;
+        }
+    }
+
+    /* Compare with the existing content of copp tables, in case values merged in helper map
+     * will be ignored as they are duplicates. In case the value in helper (merge of init and
+     * user config) will overwrie the copp table entries, current entry will be deleted from copp
+     * tables.
+     */
+    for (auto i : helper)
+    {
+        string table_key = i.first;
+        std::vector<FieldValueTuple> helper_fvs = i.second;
+        std::vector<FieldValueTuple> preserved_fvs;
+        std::vector<FieldValueTuple> new_fvs;
+
+        bool key_found = m_coppTable.get(table_key, preserved_fvs);
+        if (!key_found)
+        {
+            // key is not found in preserved copp table -> add
+            m_cfg[table_key] = helper[table_key];
+        }
+        else
+        {
+            unordered_map<string, string> preserved_copp_entry;
+            for (auto prev_fv : preserved_fvs)
+            {
+                preserved_copp_entry[fvField(prev_fv)] = fvValue(prev_fv);
+            }
+            for (auto helper_fv: helper_fvs)
+            {
+                string field = fvField(helper_fv);
+                string value = fvValue(helper_fv);
+                unordered_map<string, string>::const_iterator preserved_copp_it = preserved_copp_entry.find(field);
+                bool field_found = (preserved_copp_it != preserved_copp_entry.end());
+                bool overwrite = false;
+                if (field_found)
+                {
+                    if (!preserved_copp_it->second.compare(value))
+                    {
+                        // duplicate -> ignore and continue to next key.
+                        continue;
+                    }
+                    else
+                    {
+                        // overwrite -> delete preserved FV tuple from copp table and add helper fvs to m_cfg
+                        overwrite = true;
+                        m_coppTable.hdel(table_key, field);
```

Comment:
I don't think there is logic in copporch to process hdel. It should be del and recreating the key

Reply:
I manually checked it, this worked well in the following scenario:

1. redis-cli -n 0 hgetall "COPP_TABLE:queue4_group2" (to see the current values)
2. redis-cli -n 0 hdel "COPP_TABLE:queue4_group2" "cbs"
3. redis-cli -n 0 hset "COPP_TABLE:queue4_group2" "cbs" "700"
4. perform fast-reboot
5. copp tables are preserved through db migration
6. redis-cli -n 0 hgetall "COPP_TABLE:queue4_group2" (as soon as platform is up to check that cbs still has 700 as value)
7. redis-cli -n 0 hgetall "COPP_TABLE:queue4_group2" (sometime after, 40~ seconds to check that cbs value was changed to 600 according to the default initialization file /etc/sonic/copp_cfg.json)